h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostUtilsTest.java

```diff
@@ -56,5 +174,338 @@ public void parseFeatureScores() throws IOException, ParseException {
+    rowHeaders[arr_idx][arr_pos] = pos;
+    assertEquals(dim, pos);
+
+    final DMatrix dMatrix = new DMatrix(rowHeaders, colIndices, values, DMatrix.SparseType.CSR, dim, dim + 1, dim);
+
+    return new Matrices(dMatrix, testFrameBuilder.build());
+  }
+
+  private static double[] genIdentityMatrixFrameCol(final int colIdx, final int len) {
+    final double[] column = new double[len];
+    column[colIdx] = 1;
+    return column;
+  }
+
+
+  protected static class Matrices {
+    private final DMatrix _dmatrix;
+    private final Frame _h2oFrame;
+
+    public Matrices(DMatrix dmatrix, Frame h2oFrame) {
+      _dmatrix = dmatrix;
+      _h2oFrame = h2oFrame;
+    }
+  }
+
+  private static CsrLayout createLayout(long size, int maxArrLen) {
+    CsrLayout l = new CsrLayout();
+    l._numRegRows = (int) (size / maxArrLen);
+    l._regRowLen = maxArrLen;
+    l._lastRowLen = (int) (size - ((long) l._numRegRows * l._regRowLen)); // allow empty last row (easier and it shouldn't matter)
+    return l;
+  }
+
+  private static class CsrLayout {
+    int _numRegRows;
+    int _regRowLen;
+    int _lastRowLen;
+
+    long[][] allocateLong() {
+      long[][] result = new long[_numRegRows + 1][];
+      for (int i = 0; i < _numRegRows; i++) {
+        result[i] = new long[_regRowLen];
+      }
+      result[result.length - 1] = new long[_lastRowLen];
+      return result;
+    }
+
+    int[][] allocateInt() {
+      int[][] result = new int[_numRegRows + 1][];
+      for (int i = 0; i < _numRegRows; i++) {
+        result[i] = new int[_regRowLen];
+      }
+      result[result.length - 1] = new int[_lastRowLen];
+      return result;
+    }
+
+    float[][] allocateFloat() {
+      float[][] result = new float[_numRegRows + 1][];
+      for (int i = 0; i < _numRegRows; i++) {
+        result[i] = new float[_regRowLen];
+      }
+      result[result.length - 1] = new float[_lastRowLen];
+      return result;
+    }
+
+  }
+
+  /**
+   * Compares H2O XGBoost preds with native preds
+   *
+   * @param nativePreds Native predictions
+   * @param h2oPreds    H2O-provided predictions
+   * @param delta
+   */
+  private static void comparePreds(final float[][] nativePreds, final Vec h2oPreds, final float delta) {
+    if (nativePreds.length != h2oPreds.length()) {
+      throw new IllegalStateException(String.format("Predictions do not have the same length. Native: %x, H2O: %x",
+              nativePreds.length,
+              h2oPreds.length()));
+
+    }
+    for (int i = 0; i < nativePreds.length; i++) {
+      assertEquals(nativePreds[i][0], (float) h2oPreds.at(i), delta);
+    }
+  }
+
+
+  @RunWith(Parameterized.class)
+  public static final class XGBoostSparseMatrixTest extends XGBoostUtilsTest {
+
+    private static final int MAX_ARR_SIZE = Integer.MAX_VALUE - 10;
+
+    @Parameterized.Parameters
+    public static Collection<Object[]> data() {
+      return Arrays.asList(new Object[][]{
+              {30, 10, MAX_ARR_SIZE},
```

Comment:
Goal is to test native predictions trained on matrix with various layout (effectively 1D, 2D) of internal memory representation. Same applies for H2O DMatrix construction. Predictions should be the same each time.